### PR TITLE
Check task deprecation

### PR DESCRIPTION
**Pull request by DenisRumyantsev**
***Description:*** Check task deprecation in the `Initialize job` step for each task used in the pipeline.

How it looks:

* In the `Initialize job` logs:

image

* In the web page of the pipeline run:

image

* Example of the warning message when the task has the `docs.microsoft.com` help URL:

##[warning]Task 'Download pipeline artifact' version 0 (DownloadPipelineArtifact@0) is deprecated. This task will be removed. From April 22, 2024, onwards it may no longer be available. Please see https://learn.microsoft.com/azure/devops/pipelines/tasks/reference/download-pipeline-artifact-v0 for more information about this task.

### CODE REVIEW

**Review by vmapetr — APPROVED** [review]

File: src/Agent.Worker/TaskManager.cs [hunk 1]
```diff
@@ -308,6 +310,29 @@ private async Task DownloadAsync(IExecutionContext executionContext, Pipelines.T
...
+            if (deprecated != null && deprecated.Value<bool>())
+            {
+                int majorVersion = new Version(task.Version).Major;
+                string deprecationMessage = $"Task '{task.Name}' version {majorVersion} is deprecated";
+                var removalDate = taskJson["removalDate"];
+
+                if (removalDate != null)
+                {
+                    string removalDateString = removalDate.Value<DateTime>().ToString("MMMM d, yyyy");
+                    deprecationMessage += $" and will be removed on {removalDateString}";
```
Comment (merlynomsft) [inline src/Agent.Worker/TaskManager.cs:329]: >and will be removed on [](http://example.com/codeflow?start=45&length=22)

I'd adjust to "may not be available after {removalDateString}. "  Also should contain link for more info e.g. aka.ms/ado/taskremoval

**Comment by merlynomsft** [comment]
Adding @geekzter for review

File: src/Agent.Worker/TaskManager.cs (hunk 1 above)
Comment (DenisRumyantsev) [reply src/Agent.Worker/TaskManager.cs]: Added the help URL for tasks which have the help URL `https://docs.microsoft.com/azure/devops/pipelines/tasks/<category>/<task-name>`

File: src/Agent.Sdk/Knob/AgentKnobs.cs [hunk 2]
```diff
@@ -512,11 +512,17 @@ public class AgentKnobs
...
             "Ignores the VSTSTaskLib folder when copying tasks.",
             new RuntimeKnobSource("AZP_AGENT_IGNORE_VSTSTASKLIB"),
             new EnvironmentKnobSource("AZP_AGENT_IGNORE_VSTSTASKLIB"),
             new BuiltInDefaultKnobSource("false"));
+
+        public static readonly Knob DisableTaskDeprecationCheck = new Knob(
+            nameof(DisableTaskDeprecationCheck),
+            "Disables the task deprecation check.",
+            new EnvironmentKnobSource("AZP_AGENT_DISABLE_TASK_DEPRECATION_CHECK"),
+            new BuiltInDefaultKnobSource("false"));
```
Comment (kirill-ivlev) [inline src/Agent.Sdk/Knob/AgentKnobs.cs:526]: I think, we should reverse it to "AZP_AGENT_CHECK_FOR_TASK_DEPRICATION"

**Review by kirill-ivlev — CHANGES_REQUESTED** [review]

File: src/Agent.Worker/TaskManager.cs [hunk 3]
```diff
@@ -308,6 +315,44 @@ private async Task DownloadAsync(IExecutionContext executionContext, Pipelines.T
...
+            {
+                string friendlyName = taskJson["friendlyName"].Value<string>();
+                int majorVersion = new Version(task.Version).Major;
+                string deprecationMessage = $"Task '{friendlyName}' version {majorVersion} ({task.Name}@{majorVersion}) is deprecated";
+                var removalDate = taskJson["removalDate"];
+
+                if (removalDate != null)
+                {
+                    string removalDateString = removalDate.Value<DateTime>().ToString("MMMM d, yyyy");
+                    deprecationMessage += $" and will be removed. From {removalDateString} onwards it may no longer be available";
```
Comment (merlynomsft) [inline src/Agent.Worker/TaskManager.cs:335]: Should these strings should be localized?

https://github.com/microsoft/azure-pipelines-agent/pull/4454

**Review by kirill-ivlev — APPROVED** [review]

File: .azure-pipelines/build-job.yml [hunk 4]
```diff
@@ -138,6 +138,7 @@ jobs:
     - script: ${{ variables.devCommand }} testl0 Debug ${{ parameters.os }}-${{ parameters.arch }}
       workingDirectory: src
       displayName: Unit tests
+      timeoutInMinutes: 5
```
Comment (merlynomsft) [inline .azure-pipelines/build-job.yml:141]: Is this change needed?

File: src/Agent.Sdk/Knob/AgentKnobs.cs [hunk 5]
```diff
@@ -512,11 +512,17 @@ public class AgentKnobs
...
             new RuntimeKnobSource("AZP_AGENT_IGNORE_VSTSTASKLIB"),
             new EnvironmentKnobSource("AZP_AGENT_IGNORE_VSTSTASKLIB"),
             new BuiltInDefaultKnobSource("false"));
+
+        public static readonly Knob CheckForTaskDeprecation = new Knob(
+            nameof(CheckForTaskDeprecation),
+            "If true, the agent will check in the 'Initialize job' step each task used in the job for task deprecation.",
+            new EnvironmentKnobSource("AZP_AGENT_CHECK_FOR_TASK_DEPRECATION"),
+            new BuiltInDefaultKnobSource("false"));
     }
```
Comment (merlynomsft) [inline src/Agent.Sdk/Knob/AgentKnobs.cs:527]: RuntimeKnobSource needed for feature flag?

File: src/Misc/layoutbin/en-US/strings.json [hunk 6]
```diff
@@ -262,6 +262,9 @@
   "DeploymentPoolNotFound": "Deployment pool not found: '{0}'",
   "DeprecatedNode6": "This task uses Node 6 execution handler, which will be removed March 31st 2022. If you are the developer of the task - please consider the migration guideline to Node 10 handler - https://aka.ms/migrateTaskNode10 (check this page also if you would like to disable Node 6 deprecation warnings). If you are the user - feel free to reach out to the owners of this task to proceed on migration.",
   "DeprecatedRunner": "Task '{0}' is dependent on a task runner that is end-of-life and will be removed in the future. Authors should review Node upgrade guidance: https://aka.ms/node-runner-guidance.",
+  "DeprecationMessage": "Task '{0}' version {1} ({2}@{1}) is deprecated.",
+  "DeprecationMessageHelpUrl": "Please see {0} for more information about this task.",
+  "DeprecationMessageRemovalDate": "This task will be removed. From {0}, onwards it may no longer be available.",
```
Comment (merlynomsft) [inline src/Misc/layoutbin/en-US/strings.json:267]: Suggestion: I don't think we need "This task will be removed".  

I'm thinking: 
"Task X version Y is deprecated.  From {0}, onwards it may no longer be available" 

is sufficient.

@geekzter

**Review by merlynomsft — APPROVED** [review]

File: .azure-pipelines/build-job.yml (hunk 4 above)
Comment (DenisRumyantsev) [reply .azure-pipelines/build-job.yml:141]: Yes, either unit tests pass quickly, or they hang for an hour and fall.

File: src/Agent.Sdk/Knob/AgentKnobs.cs (hunk 5 above)
Comment (DenisRumyantsev) [reply src/Agent.Sdk/Knob/AgentKnobs.cs:527]: `RuntimeKnobSource` is not needed for FF since it is `false` by default and will be enabled on the ADO server side via the TFS CC release.

File: src/Misc/layoutbin/en-US/strings.json (hunk 6 above)
Comment (DenisRumyantsev) [reply src/Misc/layoutbin/en-US/strings.json:267]: According to the warning message, it should be clear that the task will definitely be removed after the specified date, but the term of the removal, which starts from the specified date, is not defined.
